### DISCUSSION TITLE
Can I create a variable that is not a column in a table but is a field in a model?

### DISCUSSION THREAD

**Discussion by recsater** [fastapi/sqlmodel#729] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class Test(SQLModel, table=True):
    id: str = Field(primary_key=True) # This is table column.
    age: int = ?? # this is not table column, but can be serialized with functions like model_dump_json.
```

I also tried https://sqlmodel.tiangolo.com/tutorial/fastapi/multiple-models/#multiple-models-with-inheritance
```python
class HeroBase(SQLModel):
    name: str = Field(primary_key=True)
    secret_name: str
    age: int | None = Field(default=None, index=True)
class Hero(HeroBase, table=True):
    id: int | None = Field(default=None)

class HeroUpdate(HeroBase):
    pass # name, secret_name can be None. but HeroUpdate() raises validation error.
```

### Description

I'd like to retrieve data from a database and provide additional fields to the user. How can I achieve this?

thank you.

### Operating System

Linux

### SQLModel Version

0.0.14

### Python Version

3.11.6

### Additional Context

And I tried PrivateAttr as well, but that doesn't serialize either.

**Comment by wangxin688** [comment]
seems it's not support `hybrid_property` or `colunm_property`  in `sqlmodel` yet. I guess that's you want in sqlalchemy.

**Reply by PipeKnight** [reply]
Well, I'm using `column_property` with SQLModel without any issues. Like this:

```python
QuestionPool.questions_count = column_property(
    select(func.count(LinkPoolQuestion.question_version_id))
    .where(LinkPoolQuestion.question_pool_id == QuestionPool.id)
    .correlate(QuestionPool)
    .scalar_subquery(),
)
```

**Reply by PipeKnight** [reply]
I checked and `hybrid_property` could be used in the same way. Described here:
https://github.com/tiangolo/sqlmodel/issues/299#issuecomment-1785578554

**Reply by Jijun-TANG** [reply]
@PipeKnight Thanks so much for the reply, it helps a lot!

**Comment by PipeKnight** [comment]
See my reply here: https://github.com/tiangolo/sqlmodel/discussions/729#discussioncomment-7862685
Both `column_property` and `hybrid_property` are supported is you're using them straight-forward and not like decorators

**Reply by recsater** [reply]
I still can't create a serializable variable that isn't a column in a table.

**Comment by seedor87** [comment]
Also cannot figure this out. Was looking for a workaround for a deficiency of Relationship fields, but without this the pydantic model and the sa model need to remain broken apart. It would be very nice to have this added as a feature, or at least more clearly documented.

**Comment by redb0** [comment]
@recsater

It's not entirely clear what you want to get as a result.

In your example, the `name` field cannot be `None`, it is explicitly annotated as `str`. To make it optional, you need to make an annotation `foo: str | None = Field(default=None)`

**Reply by seedor87** [reply]
SQLModel will require a column for the Field called `foo` to be created/exist in the table for HeroBase. This is not what is desired by op.
I am almost sure that using `default=None` will NOT have the effect of making the validation of the db table structure for the column `foo` elective.
EDIT: I just looked into the SQLModel Field code a little closer and it appears to relay the value directly to pydantic's Field, 
> default: Default value if the field is not set.

**Reply by redb0** [reply]
@seedor87 

If I understand correctly, you want to make a field that will not be in the database, but it will be in the model. You will have to make an additional model with this field. But then you will have to fill in this field manually

```python
from typing import Optional
from sqlmodel import Field, SQLModel


class Base(SQLModel):
    name: str = Field(
        max_length=100,
        description='Name',
    )


class Table(Base):
    id: Optional[int] = Field(
        default=None,
        primary_key=True,
        description='ID',
    )

class TablePublic(Base):
    age: Optional[int] = Field(
        default=None,
    )


a = Table(id=1, name='Foo')

b = TablePublic.model_validate(a)
print(f'{b = }')  # <-- TablePublic(name='Foo', age=None)

c = TablePublic.model_validate(a, update={'age': 42})
print(f'{c = }')  # <-- TablePublic(name='Foo', age=42)
```

**Reply by rt87** [reply]
Same problem...

However, if I declare an attribute as private (e.g. by a leading underscore), it will NOT be considered a column, i.e. it is a mere Python object attribute, but no table column.

**Comment by rohitsodhia** [comment]
I hope it's ok if I follow up on this. I'm interested in SQLModel, namely as I'm learning FastAPI, but one thing that isn't clear/a feature I'd love is the ability to mark a field in the model as not part of the table. I did see above the explained mechanism of using inheritance to differentiate models, but that feels like overkill if all we're doing is having fields that may be used in code, but aren't saved to the database, or a constant. In these cases, it wouldn't be Pythonic to add preceding underscores, and in the case of a constant, it'd be all uppercase. In SQLAlchemy, we just wouldn't annotate it with `Mapped`. How can I do this with SQLModel instead? The docs say `Only inherit from data models`, but this is the opposite, where we want more fields on the class that will be used than the table has.

**Comment by WilliamQue** [comment]
Almost one year has passed, is there a solution coming out?

**Reply by tkdrob** [reply]
Having you tried typing an attribute with ClassVar? If you are using fastapi, uou can also define a property in the model and then include that property name in a response model.

age: `ClassVar[int]`

**Comment by wangxin688** [comment]
I don't think dynamic fields serialization are related to sqlmodel or sqlalchemy, the `hybrid_property` `column_property` or pydantic `computed_filed` are meet the requirements for your needs to define a field not in table. but you need to write your own serialization logic manually. 

something like this:

```python
from pydantic import BaseModel, Field, computed_field

class MyModel(BaseModel):
    field1: str
    field2: int

    @computed_field
    @property
    def dynamic_field(self) -> str:
        return f"{self.field1}-{self.field2}"

    def to_dict(self) -> dict:
        data = self.model_dump()
        data['dynamic_field'] = self.dynamic_field
        return data
```
